File: services/hr-core/app/services/application_service.py
from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.models.application import Application, ApplicationStage
from app.models.vacancy import Vacancy
from app.schemas.application import ApplicationCreate
from app.schemas.screening import ScreeningRequest
from app.services.llm_client import LLMClient
from app.services.notifier import send_clarifying_questions


class ApplicationService:
    def __init__(self, llm_client: LLMClient | None = None) -> None:
        self.llm_client = llm_client or LLMClient()

    def create_application(self, db: Session, payload: ApplicationCreate) -> Application:
        vacancy = db.get(Vacancy, payload.vacancy_id)
        if vacancy is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Vacancy not found")

        application = Application(
            vacancy_id=payload.vacancy_id,
            candidate_email=str(payload.candidate_email),
            resume_text=payload.resume_text,
            stage=ApplicationStage.NEW.value,
        )
        db.add(application)
        db.flush()

        screening = self.llm_client.screen_resume(
            ScreeningRequest(
                resume_text=payload.resume_text,
                mandatory_requirements=vacancy.mandatory_requirements,
                optional_requirements=vacancy.optional_requirements,
            )
        )

        application.clarifying_questions = screening.clarifying_questions
        application.score = screening.score
        application.score_reasons = screening.score_reasons
        application.risks_to_clarify = screening.risks_to_clarify
        application.stage = ApplicationStage.QUESTIONS_SENT.value

        send_clarifying_questions(str(payload.candidate_email), screening.clarifying_questions)

        db.commit()
        db.refresh(application)
        return application

File: services/hr-core/src/services/applicationService.ts
import { AppError } from "../domain/errors.js";
import type { Application } from "../domain/types.js";
import type { Repository } from "../infra/repository.js";
import type { LLMClient } from "./llmClient.js";

export class ApplicationService {
  constructor(
    private readonly repo: Repository,
    private readonly llmClient: LLMClient
  ) {}

  async createApplication(input: {
    vacancyId: string;
    candidateEmail: string;
    resumeText: string;
  }): Promise<Application> {
    const vacancy = await this.repo.findVacancyById(input.vacancyId);
    if (!vacancy) {
      throw new AppError(404, "Vacancy not found");
    }

    const screening = await this.llmClient.prepareScreening({
      vacancy,
      resumeText: input.resumeText
    });

    return await this.repo.createApplication({
      vacancyId: input.vacancyId,
      candidateEmail: input.candidateEmail,
      stage: "questions_sent",
      resumeText: input.resumeText,
      answers: {},
      candidateProfile: screening.candidateProfile,
      clarifyingQuestions: screening.clarifyingQuestions,
      score: null,
      scoreReasons: [],
      risksToClarify: []
    });
  }

  async submitAnswers(applicationId: string, answers: Record<string, string>): Promise<Application> {
    const application = await this.repo.findApplicationById(applicationId);
    if (!application) {
      throw new AppError(404, "Application not found");
    }

    const vacancy = await this.repo.findVacancyById(application.vacancyId);
    if (!vacancy) {
      throw new AppError(404, "Vacancy not found");
    }

    const clarifyingQuestions = application.clarifyingQuestions.map((question, index) => ({
      id: `q${index + 1}`,
      text: question
    }));

    const ranking = await this.llmClient.rankCandidate({
      vacancy,
      candidateProfile: application.candidateProfile,
      clarifyingQuestions,
      answers
    });

    const updated = await this.repo.updateApplication(applicationId, {
      answers,
      stage: "in_review",
      score: ranking.score,
      scoreReasons: ranking.scoreReasons,
      risksToClarify: ranking.risksToClarify
    });

    if (!updated) {
      throw new AppError(404, "Application not found");
    }

    return updated;
  }
}
